File: benchmark.py
import pyarrow as pa
import pyarrow.parquet as pq
import time
import sys
import os

def process_parquet_file(file_path):
    start = time.time()

    # Open the Parquet file
    open_start = time.time()
    try:
        parquet_file = pq.ParquetFile(file_path)
    except Exception as e:
        print(f"Error opening file: {e}")
        return None, None, None
    open_end = time.time()
    open_duration = open_end - open_start

    # Read the table
    read_start = time.time()
    try:
        table = parquet_file.read()
    except Exception as e:
        print(f"Error reading file: {e}")
        return None, None, None
    read_end = time.time()
    read_duration = read_end - read_start

    end = time.time()
    total_duration = end - start

    return open_duration, read_duration, total_duration

def main():
    if len(sys.argv) < 2 or len(sys.argv) > 3:
        print(f"Usage: {sys.argv[0]} <parquet_file_path> [iterations]")
        return 1

    file_path = sys.argv[1]
    iterations = int(sys.argv[2]) if len(sys.argv) == 3 else 10


    # Set Arrow to use only one thread
    os.environ["OMP_NUM_THREADS"] = "1"
    os.environ["ARROW_NUM_THREADS"] = "1"

    print(f"Number of CPU cores available: {os.cpu_count()}")
    print(f"Number of threads used by Arrow: {pa.cpu_count()}")

    total_open_duration = 0.0
    total_read_duration = 0.0
    total_duration = 0.0

    valid_iterations = 0

    for _ in range(iterations):
        open_duration, read_duration, duration = process_parquet_file(file_path)
        if open_duration is not None:
            total_open_duration += open_duration
            total_read_duration += read_duration
            total_duration += duration
            valid_iterations += 1

    if valid_iterations > 0:
        average_open_duration = total_open_duration / valid_iterations
        average_read_duration = total_read_duration / valid_iterations
        average_duration = total_duration / valid_iterations

        print(f"Average time to open file: {average_open_duration} seconds")
        print(f"Average time to read table: {average_read_duration} seconds")
        print(f"Average total time: {average_duration} seconds")
    else:
        print("No valid iterations to calculate averages.")

if __name__ == "__main__":
    main()

File: benchmark.cc
#include <arrow/api.h>
#include <arrow/io/file.h>
#include <parquet/arrow/reader.h>
#include <iostream>
#include <memory>
#include <chrono>
#include <arrow/util/cpu_info.h>
#include <arrow/util/thread_pool.h>

arrow::Status ProcessParquetFile(const std::string& file_path, double& total_open_duration, double& total_read_duration, double& total_duration) {
    auto start = std::chrono::high_resolution_clock::now();
    arrow::MemoryPool* pool;
    ARROW_RETURN_NOT_OK(mimalloc_memory_pool(&pool));

    std::shared_ptr<arrow::io::RandomAccessFile> input;
    ARROW_ASSIGN_OR_RAISE(input, arrow::io::ReadableFile::Open(file_path));

    parquet::ArrowReaderProperties arrow_reader_properties = parquet::default_arrow_reader_properties();
    arrow_reader_properties.set_pre_buffer(true);
    arrow_reader_properties.set_use_threads(true);

    parquet::ReaderProperties reader_properties = parquet::ReaderProperties(pool);

    auto open_start = std::chrono::high_resolution_clock::now();
    std::unique_ptr<parquet::arrow::FileReader> arrow_reader;
    parquet::arrow::FileReaderBuilder reader_builder;
    reader_builder.properties(arrow_reader_properties);

    ARROW_RETURN_NOT_OK(reader_builder.Open(input));
    ARROW_RETURN_NOT_OK(reader_builder.Build(&arrow_reader));

    auto open_end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> open_duration = open_end - open_start;
    total_open_duration += open_duration.count();

    auto read_start = std::chrono::high_resolution_clock::now();
    std::shared_ptr<arrow::Table> table;
    ARROW_RETURN_NOT_OK(arrow_reader->ReadTable(&table));
    auto read_end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> read_duration = read_end - read_start;
    total_read_duration += read_duration.count();

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> duration = end - start;
    total_duration += duration.count();

    return arrow::Status::OK();
}

int main(int argc, char** argv) {
    if (argc < 2 || argc > 3) {
        std::cerr << "Usage: " << argv[0] << " <parquet_file_path> [iterations]" << std::endl;
        return 1;
    }

    std::string file_path = argv[1];

    arrow::SetCpuThreadPoolCapacity(1);

    double total_open_duration = 0.0;
    double total_read_duration = 0.0;
    double total_duration = 0.0;
    int iterations = (argc == 3) ? std::stoi(argv[2]) : 100;

    for(int i=0; i<iterations; i++) {
        arrow::Status status = ProcessParquetFile(file_path, total_open_duration, total_read_duration, total_duration);
        if (!status.ok()) {
            std::cerr << "Error: " << status.ToString() << std::endl;
            return 1;
        }
    }

    double average_open_duration = total_open_duration / iterations;
    double average_read_duration = total_read_duration / iterations;
    double average_duration = total_duration / iterations;

    std::cout << "Average time to open file: " << average_open_duration << " seconds" << std::endl;
    std::cout << "Average time to read file: " << average_read_duration << " seconds" << std::endl;
    std::cout << "Average total time: " << average_duration << " seconds" << std::endl;

    int num_cores = arrow::internal::CpuInfo::GetInstance()->num_cores();
    std::cout << "Number of CPU cores available: " << num_cores << std::endl;
    int num_threads = arrow::GetCpuThreadPoolCapacity();
    std::cout << "Number of threads in the Arrow thread pool: " << num_threads << std::endl;

    return 0;
}
